planner/cascades/implementation_rules.go

```diff
@@ -145,23 +149,71 @@ func (r *ImplTiKVSingleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *pr
 	return impl.NewTableReaderImpl(reader, sg.Source.TblColHists), nil
 }
 
+// ImplTiKVDoubleReadGather implements TiKVDoubleGather as PhysicalIndexLookUpReader.
+type ImplTiKVDoubleReadGather struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTiKVDoubleReadGather) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	if prop.IsEmpty() {
+		return true
+	}
+	reader := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	if all, _ := prop.AllSameOrder(); all {
+		for i, col := range reader.IndexCols {
+			if col.Equal(nil, prop.Items[0].Col) {
+				return plannercore.MatchIndicesProp(reader.IndexCols[i:], reader.IndexColLens[i:], prop.Items)
+			}
+		}
+	}
+	return false
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplTiKVDoubleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	logicProp := expr.Group.Prop
+	sg := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	var reader plannercore.PhysicalPlan
+	reader = sg.GetPhysicalIndexLookUpReader(logicProp.Schema, logicProp.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt), reqProp, reqProp)
```

Comment:
1. `reqProp` should be cloned.
2. Only the index path can keep the order?

Reply:
1. OK.
2. Yes, only the index scan needs to keep order.